### PR TITLE
land_detector: robustify land detection by using distance to ground info

### PR DESCRIPTION
**Pull request by RomanBapst**
False ground-contact detection can happen with multi rotors during a maneuver where the vehicle decelerates horizontally while the user is demanding a sink rate. In this case the vehicle can "float" for a certain amount of time which can be registered as ground-contact by the land detector.
If the vehicle is equipped with a distance sensor we can make use of that data to increase the hysteresis used for the land detection states.
If the system knows that it can measure distance to the ground and the distance is greater than 1m, then the hysteresis values will be increased by a factor of 3 and thus should drastically reduce the chances of false ground contact detection.
Notice that the distance measurement itself is only used to determine the value of the hysteresis, however, it will not be used directly for making the ground contact decision.
In other words, a distance sensor reporting false data will either cause ground contact detection to be delayed (in case it reports > 1m) or it will cause the usual hysteresis values to be used (in case it reports < 1m). But you should never get a false positive ground detection due to a faulty lidar, which is important.

Also note, that these changes should have no impact on systems which don't have a distance to ground measurement available.

### CODE REVIEW

File: src/modules/land_detector/LandDetector.cpp [hunk 1]
```diff
@@ -96,6 +96,22 @@ void LandDetector::Run()
 
 	_update_topics();
 
+	if (!_dist_bottom_is_observable) {
```
Comment (dagar) [inline src/modules/land_detector/LandDetector.cpp:99]: Once it's observable you never check again?

Comment (RomanBapst) [reply src/modules/land_detector/LandDetector.cpp:99]: @dagar Yes, the assumption is that the configuration will not change. Actually, I currently don't see any reason to not make it update.

Comment (RomanBapst) [reply src/modules/land_detector/LandDetector.cpp:99]: @dagar Hm, on a second thought I think I know why it was implemented like this. Consider the time when a range finder stops working, the ekf will then report dist bottom not to be observable. In that case you probably still want to assume that you have a range finder but the data is not valid and thus keep a high hysteresis. It's better than switching to low hysteresis and risking a false ground detection positive.
Thoughts welcome.

**Comment by RomanBapst** [comment]
@dagar @MaEtUgR Any more thoughts on this. This should considerably reduce risk of false land detection when you have a lidar available so I think it's quite important.

**Comment by RomanBapst** [comment]
@dagar @MaEtUgR @bkueng Any reason not to merge this?

**Comment by dagar** [comment]
> @dagar @MaEtUgR @bkueng Any reason not to merge this?

Seems good to me unless we need to reconcile with more recent discovered problems.

**Comment by RomanBapst** [comment]
@dagar @MaEtUgR I need somebody to approve.

**Review by dagar — APPROVED** [review]

File: src/modules/land_detector/MulticopterLandDetector.cpp [hunk 2]
```diff
@@ -355,4 +357,22 @@ bool MulticopterLandDetector::_get_ground_effect_state()
...
+		return false;
+	}
+}
+
+void MulticopterLandDetector::_set_hysteresis_factor(const int factor)
+{
+	_ground_contact_hysteresis.set_hysteresis_time_from(false, GROUND_CONTACT_TRIGGER_TIME_US * factor);
+	_landed_hysteresis.set_hysteresis_time_from(false, LAND_DETECTOR_TRIGGER_TIME_US * factor);
+	_maybe_landed_hysteresis.set_hysteresis_time_from(false, MAYBE_LAND_DETECTOR_TRIGGER_TIME_US * factor);
+	_freefall_hysteresis.set_hysteresis_time_from(false, FREEFALL_TRIGGER_TIME_US);
```
Comment (MaEtUgR) [inline src/modules/land_detector/MulticopterLandDetector.cpp:375]: @RomanBapst Why is a freefall less likely to happen if you're close to the ground? :thinking:

Comment (MaEtUgR) [reply src/modules/land_detector/MulticopterLandDetector.cpp:375]: Ha, it's not but it's still in the hysteresis factor function... Nevermind. Implementation detail.
